doc/source/structures/misc/string.rst

```diff
@@ -326,6 +329,48 @@ Structure
 
     returns a new string with no leading whitespace
 
+.. method:: String:TONUMBER(defaultIfError)
+
+    :parameter defaultIfError: (optional argument) :struct:`Scalar` to return as a default value if the string format is in error.
+    :return: :struct:`Scalar`
+
+    Returns the numeric version of the string, as a number that can be used
+    for mathematics or anywhere a :struct:`Scalar` is expected.  If the
+    string is not in a format that kOS is able to convert into a number, then
+    the value ``defaultIfError`` is returned instead.  You can use this to
+    either select a sane default, or to deliberately select a value you
+    never expect to get in normal circumstances so you can use it as a
+    test to see if the string was formatted well.
+
+    The argument ``defaultIfError`` is optional.  If it is left off, then
+    when there is a problem in the format of the string, you will get
+    an error that stops the script instead of returning a value.
+
+    The valid understood format allows an optional leading sign,
+    a decimal point with fractional part, and scientific notation
+    using "e" as in "1.23e3" for "1230" or "1.23e-3" for "0.00123".
```

Comment:
How about removing the extra step?  We just make DOUBLE parse so it allows the 'e' inside it, as follows:
```
INTEGER   ->   @"[0-9]+(e[+-]?[0-9]+)?";   // allows whole-number sci notation like "123e4"
DOUBLE    ->    @"[0-9]*\.[0-9]+(e[+-]?[0-9]+)?";   // allows sci notation with a fraction like 1.23e4"
...

// Maybe merge these two together in the syntax everywhere since one is just the same as the other now?
number   -> (DOUBLE | INTEGER);
sci_number   -> number;

```
Then VisitSciNumber() would just call TryParse for either int or double depending on which it is.  NOTICE:  the rule above would allow "123e3" to be called an INTEGER even though I don't int.TryParse  cab;t understands scientific notation, so we'd need to check that case and divert such "integers" over into being treated as doubles just to make the TryParse work.

(My reason for proposing it this way is that the regular expression for DOUBLE gets very ugly if we want to support the logic "a period is mandatory *except when* there is scientific notation, then it's optional" in it.  It's easier to just temporarily treat the period-less case as INTEGER and clean it up later in Compiler.cs.)